x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java

```diff
@@ -125,22 +129,64 @@ public void resolveAsMergedMappingAndRetrieveMinimumVersion(
-                        // we don't have that information and need to assume the oldest supported version.
-                        minimumVersion == null ? TransportVersion.minimumCompatible() : minimumVersion
-                    )
+                FieldsInfo info = new FieldsInfo(
+                    response.caps(),
+                    response.minTransportVersion(),
+                    Build.current().isSnapshot(),
+                    supportsAggregateMetricDouble,
+                    supportsDenseVector
                 );
+                LOGGER.debug("minimum transport version {} {}", response.minTransportVersion(), info.effectiveMinTransportVersion());
+                l.onResponse(new Versioned<>(mergedMappings(indexWildcard, info), info.effectiveMinTransportVersion()));
             })
         );
     }
 
-    public record FieldsInfo(FieldCapabilitiesResponse caps, boolean supportAggregateMetricDouble, boolean supportDenseVector) {}
+    /**
+     * Information for resolving a field.
+     * @param caps {@link FieldCapabilitiesResponse} from all indices involved in the query
+     * @param minTransportVersion the minimum {@link TransportVersion} of any node that <strong>might</strong> receive the request
+     * @param currentBuildIsSnapshot is the current build a snapshot? Note: This is always {@code Build.current().isSnapshot()} in
+     *                               production but tests need more control
+     * @param useAggregateMetricDoubleWhenNotSupported does the query itself force us to use {@code aggregate_metric_double} fields
+     *                                                 even if the remotes don't report that they support the type? This exists because
+     *                                                 some remotes <strong>do</strong> support {@code aggregate_metric_double} without
+     *                                                 reporting that they do. And, for a while, we used the query itself to opt into
+     *                                                 reading these fields.
+     * @param useDenseVectorWhenNotSupported does the query itself force us to use {@code dense_vector} fields even if the remotes don't
+     *                                       report that they support the type? This exists because some remotes <strong>do</strong>
+     *                                       support {@code dense_vector} without reporting that they do. And, for a while, we used the
+     *                                       query itself to opt into reading these fields.
+     */
+    public record FieldsInfo(
+        FieldCapabilitiesResponse caps,
+        @Nullable TransportVersion minTransportVersion,
+        boolean currentBuildIsSnapshot,
+        boolean useAggregateMetricDoubleWhenNotSupported,
+        boolean useDenseVectorWhenNotSupported
+    ) {
+        /**
+         * The {@link #minTransportVersion}, but if any remote didn't tell us the version we assume
```

Comment:
Can we say that, if a remote (cluster) didn't tell us the min version, it's at most on 9.1, because if all nodes where on 9.2.0+, the min transport version would be part of the field caps response?